### PR TITLE
#524 : Support for Subscriptions

### PR DESCRIPTION
**Pull request by anandrgitnirman**
#524

### CODE REVIEW

File: blockchain/serviceMetadata.go [hunk 1]
```diff
@@ -36,8 +36,87 @@ import (
             "endpoints": [
                 "https://tz-services-1.snet.sh:8005"
             ],
-            "group_id": "EoFmN3nvaXpf6ew8jJbIPVghE5NXfYupFF7PkRmVyGQ="
+            "group_id": "EoFmN3nvaXpf6ew8jJbIPVghE5NXfYupFF7PkRmVyGQ=",
+{
+  "licenses": {
+    "tiers": [{
```
Comment (ksridharbabuus) [inline blockchain/serviceMetadata.go:42]: Is having tiers & subscription as a separate element in the licenses makes it difficult to parse.
Shall we look at keeping in a generic way:
licenses: [{
type: "tiers",
details: []
},
{
type: "subscription",
details: []
},
]

File: blockchain/serviceMetadata.go [hunk 2]
```diff
@@ -36,8 +36,87 @@ import (
...
+          "planName": "Monthly For ServiceA/MethodA",
+          "LicenseCost": 90,
+          "grpcServiceName": "ServiceA",
+          "grpcMethodName": "MethodA"
+        },
+        {
+          "periodInDays": 30,
+          "creditsInAGI": 123,
+          "planName": "Monthly",
+          "LicenseCost": 93
```
Comment (ksridharbabuus) [inline blockchain/serviceMetadata.go:99]: Why the LicenseCost field name different naming convention?

File: blockchain/serviceMetadata.go [hunk 3]
```diff
@@ -76,13 +155,33 @@ type ServiceMetadata struct {
 	freeCallsAllowed      int
 }
 
+type Subscription struct {
+	PeriodInDays    int    `json:"periodInDays"`
+	CreditsInAGI    int    `json:"creditsInAGI"`
+	PlanName        string `json:"planName"`
+	LicenseCost     int    `json:"LicenseCost"`
```
Comment (ksridharbabuus) [inline blockchain/serviceMetadata.go:162]: Need to check LicenseCost here as well...

File: blockchain/serviceMetadata.go [hunk 4]
```diff
@@ -36,8 +36,87 @@ import (
...
+        {
+          "high": 200,
+          "fixedPriceInCogs": 0.75
+        },
+        {
+          "high": 300,
+          "fixedPriceInCogs": 0.50
         }
+      ],
+      "detailsUrl": "http://abc.org/licenses/Tier.html",
```
Comment (raamb) [inline blockchain/serviceMetadata.go:81]: Ensure that this can be a https or ipfs hash

File: blockchain/serviceMetadata_test.go [hunk 5]
```diff
@@ -10,11 +11,12 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+var testLicenseJsonData = "  \"licenses\": {\n        \"subscriptions\": {\n   \"subscription\": [\n  {\n    \"periodInDays\": 30,\n    \"creditsInAGI\": 120,\n    \"planName\": \"Monthly For ServiceA/MethodA\",\n    \"LicenseCost\": 90,\n    \"grpcServiceName\": \"ServiceA\",\n    \"grpcMethodName\": \"MethodA\"\n  },\n  {\n    \"periodInDays\": 30,\n    \"creditsInAGI\": 123,\n    \"planName\": \"Monthly\",\n    \"LicenseCost\": 93\n  },\n  {\n    \"periodInDays\": 120,\n    \"creditsInAGI\": 160,\n    \"LicenseCost\": 120,\n    \"planName\": \"Quarterly\"\n  },\n  {\n    \"periodInDays\": 365,\n    \"creditsInAGI\": 430,\n    \"LicenseCost\": 390,\n    \"planName\": \"Yearly\"\n  }\n],       \"type\": \"Subscription\",\n          \"detailsUrl\": \"http://abc.org/licenses/Subscription.html\",\n          \"isActive\": \"true/false\"\n        }\n      }"
```
Comment (raamb) [inline blockchain/serviceMetadata_test.go:14]: Please add a testcase for service level pricing (i.e. no price at method level)

File: blockchain/serviceMetadata.go (hunk 1 above)
Comment (anandrgitnirman) [reply blockchain/serviceMetadata.go:42]: the below is what we had last revivewed , keeping them as seprate constructs was actually easier to parse, the attributes of these license types are very differenet mostly,
```json
"tier": {
      "type": "Tier",
      "planName": "Tier BBB Applicable for All service.methods",
      "range": [
        {
          "high": 100,
          "fixedPriceInCogs": 1
        },
        {
          "high": 200,
          "fixedPriceInCogs": 0.75
        },
        {
          "high": 300,
          "fixedPriceInCogs": 0.50
        }
      ],
      "detailsUrl": "http://abc.org/licenses/Tier.html",
      "isActive": "true/false"
    },

"subscriptions": {
      "type": "Subscription",
      "periods": [
        {
          "periodInDays": 30,
          "creditsInAGI": 120,
          "planName": "Monthly For ServiceA/MethodA",
          "LicenseCost": 90,
          "grpcServiceName": "ServiceA",
          "grpcMethodName": "MethodA"
        },
        {
          "periodInDays": 30,
          "creditsInAGI": 123,
          "planName": "Monthly",
          "LicenseCost": 93
        },
````

File: blockchain/serviceMetadata.go (hunk 4 above)
Comment (anandrgitnirman) [reply blockchain/serviceMetadata.go:81]: you mean the details url  ? this is something that Dapp  / other consumer would use , Daemon doesnt really use this  detailsUrl

File: blockchain/serviceMetadata.go (hunk 2 above)
Comment (anandrgitnirman) [reply blockchain/serviceMetadata.go:99]: @corrected this Sir

File: blockchain/serviceMetadata.go (hunk 3 above)
Comment (anandrgitnirman) [reply blockchain/serviceMetadata.go:162]: Corrected this Sir

**Review by ksridharbabuus — APPROVED** [review]
Approved.

**Pull request merged**